File: frontends/Matlab/master/validate_mkmaster.cpp
#include <string>
#include <fstream>
#include <iostream>
#include <vector>
#include <math.h>
#include <stdlib.h>

using namespace std;

#include "mkmaster.h"
typedef int Int;

void writedarray(ofstream &out, vector<double> &a, Int N)
{
    out.write( reinterpret_cast<char*>( &a[0] ), sizeof(double) * N );
}

template <typename T> string NumberToString ( T Number )
{
    ostringstream ss;
    ss << Number;
    return ss.str();
}

int main(int argc, char** argv) 
{
    Int quadType[] = {0,0,0};
    Int nodetype = 0;
    masterstruct master;
    
    for (Int porder=0; porder <= 6; porder++)
        for (Int quadOrder=1; quadOrder <= 3; quadOrder++)
            for (Int elemtype=0; elemtype <= 1; elemtype++)
                for (Int nd=1; nd <= 3; nd++) {
                    if (quadOrder*porder <= 16) {
                        Int pgauss[] = {quadOrder*porder,quadOrder*porder,quadOrder*porder};
                        
                        mkmaster(master, porder, &pgauss[0], &quadType[0], nd, elemtype, nodetype);
                        
                        string filename = "P" + NumberToString(porder) + "D" + NumberToString(nd) +"E" + NumberToString(elemtype) + "Q" + NumberToString(quadOrder) + ".bin";
                        ofstream out(filename.c_str(), ios::out | ios::binary);
                        
                        if (!out)
                            exit(-1);

                        writedarray(out, master.plocvl, (Int) master.plocvl.size());
                        writedarray(out, master.plocfc, (Int) master.plocfc.size());
                        writedarray(out, master.gpvlR, (Int) master.gpvlR.size());
                        writedarray(out, master.gwvlR, (Int) master.gwvlR.size());
                        writedarray(out, master.gpfcR, (Int) master.gpfcR.size());
                        writedarray(out, master.gwfcR, (Int) master.gwfcR.size());
                        writedarray(out, master.shapvlR, (Int) master.shapvlR.size());
                        writedarray(out, master.shapfcR, (Int) master.shapfcR.size());

                        out.close();
                    }
                }
    return 0;
}

File: frontends/Matlab/master/validate_mkmaster.m

clc

nodetype = 0;

for nd=1:3
    for porder=0:6
        for elemtype=0:1
            for quadOrder=1:3
                if quadOrder*porder <= 16
                    pgauss = quadOrder*porder;
                    
                    mesh.nd = nd;
                    mesh.porder = porder;
                    mesh.perm = [];
                    mesh.elemtype = elemtype;
                    [mesh.plocal,~,mesh.plocfc,~,~,~,~] = masternodes(porder,nd,elemtype,nodetype);
                    
                    master = mkmaster(mesh,pgauss);
                    
                    filename = ['P', num2str(porder), 'D', num2str(nd), 'E', num2str(elemtype), 'Q', num2str(quadOrder), '.bin'];

                    fileID = fopen(filename,'r');
                    data = fread(fileID,'double');    
                    fclose(fileID);
                    
                    start = 1;
                    N = length(master.plocvl(:));
                    endd = start + N - 1;
                    plocvl = data(start:endd);
                    errorNorm = norm(plocvl(:) - master.plocvl(:));
                    if errorNorm > 1.0e-9; disp([filename,' plocvl']); end
                        
                    start = endd + 1;
                    N = length(master.plocfc(:));
                    endd = start + N - 1;
                    plocfc = data(start:endd);
                    errorNorm = norm(plocfc(:) - master.plocfc(:));
                    if errorNorm > 1.0e-9; disp([filename,' plocfc']); end
                    
                    start = endd + 1;
                    N = length(master.gpvl(:));
                    endd = start + N - 1;
                    gpvlR = data(start:endd);
                    errorNorm = norm(gpvlR(:) - master.gpvl(:));
                    if errorNorm > 1.0e-9; disp([filename,' gpvl']); end
                    
                    start = endd + 1;
                    N = length(master.gwvl(:));
                    endd = start + N - 1;
                    gwvlR = data(start:endd);
                    errorNorm = norm(gwvlR(:) - master.gwvl(:));
                    if errorNorm > 1.0e-9; disp([filename,' gwvl']); end
                    
                    start = endd + 1;
                    N = length(master.gpfc(:));
                    endd = start + N - 1;
                    gpfcR = data(start:endd);
                    errorNorm = norm(gpfcR(:) - master.gpfc(:));
                    if errorNorm > 1.0e-9; disp([filename,' gpfc']); end
                    
                    start = endd + 1;
                    N = length(master.gwfc(:));
                    endd = start + N - 1;
                    gwfcR = data(start:endd);
                    errorNorm = norm(gwfcR(:) - master.gwfc(:));
                    if errorNorm > 1.0e-9; disp([filename,' gwfc']); end
                    
                    start = endd + 1;
                    N = length(master.shapvl(:));
                    endd = start + N - 1;
                    shapvlR = data(start:endd);
                    errorNorm = norm(shapvlR(:) - master.shapvl(:));
                    if errorNorm > 1.0e-9
                        disp([filename,' shapvl']);
                    end
                    
                    start = endd + 1;
                    N = length(master.shapfc(:));
                    endd = start + N - 1;
                    shapfcR = data(start:endd);
                    errorNorm = norm(shapfcR(:) - master.shapfc(:));
                    if errorNorm > 1.0e-9; disp([filename,' shapfc']); end
                end
            end
        end
    end
end
    